src/ADCS.SidExtension.PolicyModule/DsUtils.cs

```diff
@@ -58,12 +58,23 @@ static class DsUtils {
         if (name.Contains("@")) {
             // upn
             Int32 splitIndex = name.IndexOf("@", StringComparison.Ordinal);
-            domainTokens = name.Substring(splitIndex + 1).Split('.');
+            //domainTokens = name.Substring(splitIndex + 1).Split('.');
+            String domainPart = name.Substring(splitIndex + 1);
+            using (var forest = Forest.GetCurrentForest())
+            {
+                foreach (Domain domain in forest.Domains)
+                {
+                    if (domain.GetDirectoryEntry().Properties["uPNSuffixes"].Contains(domainPart)) {
```

Comment:
I don't think that `uPNSuffixes` attribute is part of domain object. In fact, when you create a new UPN suffix, it is created for entire forest and stored in configuration naming context which isn't bound to any particular domain it is bound to forest root domain where configuration naming context is stored. In addition, this information can be cached in order to boost the performance. This will require additional code to support this. Let me think about this and I will address this issue. Anyways, thanks for raising this point and your efforts!

Reply:
Thank you for the Policy module! Glad I could get this on your radar.